
#Workflow:

##Target:
finetune LLama model version: on dataset: specific on area: dialog summary

##Dataset:
huggingface

##Base model:
setup Libraries\
Download LLama model\
Base model Config setup(quantization setting step is here)\
import pre-trained model\
setup tokenizer\



##Data preprocess:
  1: formatting text and summary to "prompt+input:text + output:summary"for model\
  2: tokenization, transfet text to token_ids(with attention_mask)
  3: use base model to generate 'raw' summary for feeling.

##Finetune process:
  1:trainning setup:\
    1.1: Set up QLora Moudle(actually LoraConfig, base parameters W parameters NF4 setting + quantization is in earlier base model Config)
    1.2: Set up peft_trainer
    1.3: .train()

  2:check training process:\
    trainable model parameters\
    all model parameters\
    percentage of trainable model parameters

  3: train model\

##post-training:

1:upload Peft_model from based model + Lora tuend model

2:test its ability of generating dialog summary. Compare its results with base model's.

3:compute Rouge scores of base model and fine-tuned model, and check if ft_model has better ability for summerizing.

##Important comments:

1.I don't have good local GPU, and that's main reason I put this project on colab. One biggest debug issue I constantly have is error of torch.device(None). 我在peft_model.train(cuda)时触发 torch.device(None) 的报错，debug去掉了peft_model.to('cuda') 以及原模型 device—map=‘cuda’ 改成‘auto’ 后没问题了。可能是某些子模块或者参数（比如lora参数）没有放到GPU上，引发这个 “device index None” 的错误。
然后改成device_map='auto'，Accelerate就会动为模型的不同子模块分配设备，避免设备不一致带来的问题。
也可以参考这个回复https://stackoverflow.com/questions/78008119/huggingface-transformer-train-function-throwing-device-received-an-invalid-com?utm_source=chatgpt.com

2.If you want to run this file, pls change hugging face access token with your own, and before this step you need to apply premission from
Meta Llama 3 on hugging face. The access token shown in this file is changed by randomly generated one, and you will get login error with it.

3.On github you can see modified version of this project. That evaluate generated summaries in more details.

4.关于checkpointing 主要作用在训练的 前向／反向传播中的激活保存与梯度重计算；而 use_cache 主要作用在推理阶段的 Key/Value 缓存复用，这俩不是一个阶段的事情，开启Key/Value 缓存复用不会加速训练速度。另外我看到有个帖子讨论这件事，https://discuss.huggingface.co/t/why-is-use-cache-incompatible-with-gradient-checkpointing/18811，我的理解是在训练中，通常是一次处理完整序列，模型并不逐 token 生成、累积 past key values。此时 use_cache=True 会让模型保存／返回 past_key_values，但这些缓存可能会干扰后续梯度计算或被视为“状态”而非激活，从而与 checkpointing 的重新计算逻辑冲突。

In [1]:
%pip  install accelerate peft bitsandbytes transformers trl pynvml

In [2]:
import os
from huggingface_hub import login
from datasets import load_dataset
from transformers import (
AutoModelForCausalLM,
AutoTokenizer,
BitsAndBytesConfig,
TrainingArguments,
pipeline,
logging,
)
from peft import LoraConfig
from trl import SFTTrainer
from tqdm import tqdm
import torch
import time
import pandas as pd
import numpy as np
from pprint import pprint

/usr/local/lib/python3.12/dist-packages/torch/cuda/__init__.py:63: FutureWarning: The pynvml package is deprecated. Please install nvidia-ml-py instead. If you did not install pynvml directly, please report this to the maintainers of the package that installed pynvml for you.
  import pynvml  # type: ignore[import]


Next cell is upload LLama model name on huggingface, and dialog summary dataset "neil-code/dialogsum-test". Print first 10 rows to check data content.

In [3]:
HF_TOKEN = "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"  # enter huggingface account token "hf_xxx"; 如果留空且模型受限会报权限错误
if HF_TOKEN:
  login(HF_TOKEN)

model_id = "meta-llama/Meta-Llama-3-8B-Instruct"

dataset_name = "neil-code/dialogsum-test"
dataset = load_dataset(dataset_name)

pprint(dataset['train'][:3], width=100, compact=False) # pprint is useful to see output of cell in a block rather than in a line


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


{'dialogue': ["#Person1#: Hi, Mr. Smith. I'm Doctor Hawkins. Why are you here today?\n"
              '#Person2#: I found it would be a good idea to get a check-up.\n'
              "#Person1#: Yes, well, you haven't had one for 5 years. You should have one every "
              'year.\n'
              '#Person2#: I know. I figure as long as there is nothing wrong, why go see the '
              'doctor?\n'
              '#Person1#: Well, the best way to avoid serious illnesses is to find out about them '
              'early. So try to come at least once a year for your own good.\n'
              '#Person2#: Ok.\n'
              '#Person1#: Let me see here. Your eyes and ears look fine. Take a deep breath, '
              'please. Do you smoke, Mr. Smith?\n'
              '#Person2#: Yes.\n'
              '#Person1#: Smoking is the leading cause of lung cancer and heart disease, you know. '
              'You really should quit.\n'
              "#Person2#: I've tried hundreds of time

In next cell, we build a function to format data into instruction data. instruction data is for prompt use.

Notice that this function works on every single 'context combo'(id,dialog,summary,topic) and return a (text) added into the 'context combo'.

In [4]:
def create_prompt_formats(sample):
    """
    Format various fields of the sample ('instruction','output')
    :param sample: input data
    """
    INTRO_BLURB = """Instruct: Below is an instruction that describes a task.
Write a response that appropriately completes the request."""
    INSTRUCTION_KEY = "Input: Please Summarize the below conversation."
    RESPONSE_KEY = "Output:"
    blurb = f"\n{INTRO_BLURB}"
    instruction = f"{INSTRUCTION_KEY}"
    input_context = f"{sample['dialogue']}" if sample["dialogue"] else None
    response = f"{RESPONSE_KEY}\n{sample['summary']}"
    parts = [part for part in [blurb, instruction, input_context, response] if
    part]
    formatted_prompt = "\n\n".join(parts)
    sample["text"] = formatted_prompt
    return sample

print(create_prompt_formats(dataset['train'][0])['text'])


Instruct: Below is an instruction that describes a task.
Write a response that appropriately completes the request.

Input: Please Summarize the below conversation.

#Person1#: Hi, Mr. Smith. I'm Doctor Hawkins. Why are you here today?
#Person2#: I found it would be a good idea to get a check-up.
#Person1#: Yes, well, you haven't had one for 5 years. You should have one every year.
#Person2#: I know. I figure as long as there is nothing wrong, why go see the doctor?
#Person1#: Well, the best way to avoid serious illnesses is to find out about them early. So try to come at least once a year for your own good.
#Person2#: Ok.
#Person1#: Let me see here. Your eyes and ears look fine. Take a deep breath, please. Do you smoke, Mr. Smith?
#Person2#: Yes.
#Person1#: Smoking is the leading cause of lung cancer and heart disease, you know. You really should quit.
#Person2#: I've tried hundreds of times, but I just can't seem to kick the habit.
#Person1#: Well, we have classes and some medicatio

In next cell we need to build QLoRA setting. This is because we will quantlize base weight matrix from bf16 to 4-bit, and decoder it on the fly of each training iteration. So Let's start.

In [5]:
compute_dtype = getattr(torch, "float16")

quant_config = BitsAndBytesConfig(
  load_in_4bit=True,
  bnb_4bit_quant_type="nf4",
  bnb_4bit_compute_dtype=compute_dtype,
  bnb_4bit_use_double_quant=True,
)

upload base model(here we use meta-llama/Llama-3.1-8B). We also check improvement of the size of model with our quantization_config(NF4 for base model weight parameter matrix)

In [6]:
model_id = "meta-llama/Meta-Llama-3-8B-Instruct"

original_model = AutoModelForCausalLM.from_pretrained(
model_id,
dtype=compute_dtype, #`torch_dtype` is deprecated! Use `dtype` instead!
device_map="auto",
quantization_config=quant_config
)

Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]

add tokenizer

In [7]:
tokenizer = AutoTokenizer.from_pretrained(model_id, use_fast=False,
  trust_remote_code=True,
  padding_side="left",add_eos_token=True,add_bos_token=True)#left padding to promise in generation process we put non-EOS token to right side

tokenizer.pad_token_id = tokenizer.eos_token_id #llama has special padding token, just to play safe, so we use eos token to be pading token

Tese our base model before fine-tune: inference：
1: build gen() function to create template
2: use tokenizer to decoder and


In [8]:
%%time
eval_tokenizer = tokenizer

def gen(model,p, maxlen=200, sample=True):
    toks = eval_tokenizer(p, return_tensors="pt")
    res = model.generate(**toks.to("cuda"), max_new_tokens=maxlen,do_sample=sample,num_return_sequences=1,temperature=0.5,num_beams=1,top_p=0.99,
    ).to('cpu') #multi options to generate: beam search, top k, top p. We choose top p here.##generate expects named tensors (input_ids, attention_mask, …) on the same device as the model.
    #Using .to("cuda") + ** cleanly moves and passes them.
    return eval_tokenizer.batch_decode(res,skip_special_tokens=True)


from transformers import set_seed
seed = 100
set_seed(seed) #fix the random choosing happened in gen() below
index = 1
prompt = dataset['test'][index]['dialogue']
summary = dataset['test'][index]['summary']
formatted_prompt = f"Instruct: Summarize the following conversation.\nInput: {prompt}\nOutput:\n"
res = gen(original_model,formatted_prompt,100,)
dash_line = '-'.join('' for x in range(200))
print(res[0])
print(dash_line)
output = res[0].split('Output:\n')[1]
print(dash_line)
print(f'INPUT PROMPT:\n{formatted_prompt}')
print(dash_line)
print(f'BASELINE HUMAN SUMMARY:\n{summary}\n')
print(dash_line)
print(f'MODEL GENERATION - 0-SHOT:\n{output}') # My feedback after reading the results: baseline is true summary, and 0-shot is more like keep writing the story from prompt.

base_summary = []
for i in range(0,10):
  index = i
  prompt = dataset['test'][index]['dialogue']
  summary = dataset['test'][index]['summary']
  formatted_prompt = f"Instruct: Summarize the following conversation.\nInput: {prompt}\nOutput:\n"
  #pprint(formatted_prompt)
  res = gen(original_model,formatted_prompt,100,)
  dash_line = '-'.join('' for x in range(100))
  output = res[0].split('Output:\n')[1]
  base_summary.append(output)

pprint(base_summary[0])

Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.


Instruct: Summarize the following conversation.
Input: #Person1#: Ms. Dawson, I need you to take a dictation for me.
#Person2#: Yes, sir...
#Person1#: This should go out as an intra-office memorandum to all employees by this afternoon. Are you ready?
#Person2#: Yes, sir. Go ahead.
#Person1#: Attention all staff... Effective immediately, all office communications are restricted to email correspondence and official memos. The use of Instant Message programs by employees during working hours is strictly prohibited.
#Person2#: Sir, does this apply to intra-office communications only? Or will it also restrict external communications?
#Person1#: It should apply to all communications, not only in this office between employees, but also any outside communications.
#Person2#: But sir, many employees use Instant Messaging to communicate with their clients.
#Person1#: They will just have to change their communication methods. I don't want any - one using Instant Messaging in this office. It waste

Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.


('A company-wide memo restricting all office communications to email '
 'correspondence and official memos, with a warning and potential termination '
 'for employees who violate this policy. The memo is to be distributed by 4 '
 'pm. External communications are also restricted, and employees who use '
 'Instant Messaging to communicate with clients will have to change their '
 'methods. Department heads will handle any questions regarding this new '
 'policy. The memo is to be taken as an intra-office memo. The company is '
 'taking this action to prevent time-wasting. The memo')
CPU times: user 1min 16s, sys: 153 ms, total: 1min 16s
Wall time: 1min 16s


Main stage:

We want to finetune base model now.

The first step is data preprocess. We already created new dataset[train]['text'], which combines input:\n text + output:\n summary together.
Now we use tokenizer we setup earlier, and transfer text data to token ids. (embedding from tokens to vectors is inside llama transformer model, so we don't need to do this step)

It's good data for model to learn now.

Next few cells is for data preprocess.

In [9]:
def get_max_length(model):
    conf = model.config
    max_length = None
    for length_setting in ["n_positions", "max_position_embeddings",
                           "seq_length"]:
        max_length = getattr(model.config, length_setting, None)
        if max_length:
            print(f"Found max lenth: {max_length}")
            break
    if not max_length:
        max_length = 1024
        print(f"Using default max length: {max_length}")
    return max_length

def preprocess_batch(batch, tokenizer, max_length):
    """
    Tokenizing a batch
    """
    return tokenizer(batch["text"],max_length=max_length,truncation=True,) # truncation(means cut) and padding(means add) generate tokens of each text paragraph with same length.

In [10]:
from functools import partial
def preprocess_dataset(tokenizer: AutoTokenizer, max_length: int,seed,
dataset):
    # Add prompt to each sample
    print("Preprocessing dataset...")
    dataset = dataset.map(create_prompt_formats)#function in .map() acts on dataset sample by sample
    _preprocessing_function = partial(preprocess_batch, max_length=max_length,
    tokenizer=tokenizer)#fixed input max—length and tokenizer with preprocess—batch function because too many input，I may make mistake later
    dataset = dataset.map(
    _preprocessing_function,
    batched=True,
    remove_columns=['id', 'topic', 'dialogue', 'summary'],
    )#function in .map() acts on dataset sample by sample

    # filter data to drop sample that longer than max_length,
    dataset = dataset.filter(lambda sample: len(sample["input_ids"]) <=
    max_length)
    # Shuffle data with fixed seed to keep result fixed
    dataset = dataset.shuffle(seed=seed)
    print("Preprocessing dataset done.")
    return dataset

In [11]:
# apply those tools and setups above to our dataset
max_length = get_max_length(original_model)
print(max_length)
train_dataset = preprocess_dataset(tokenizer, max_length,seed,
dataset['train'])
eval_dataset = preprocess_dataset(tokenizer, max_length,seed,
dataset['validation']) #validation dataset

Found max lenth: 8192
8192
Preprocessing dataset...
Preprocessing dataset done.
Preprocessing dataset...


Map:   0%|          | 0/499 [00:00<?, ? examples/s]

Filter:   0%|          | 0/499 [00:00<?, ? examples/s]

Preprocessing dataset done.


In second step, we build a finetune model. There are two parts: first part is setup of Lora and second is finetune model. Notice that we basicly just work on parameters of LoraConfig and transformers.trainer. We don't need to build these two moudle, but it's necessary to know the concepts and structure of these two.

In [12]:
from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training

config = LoraConfig(
r=16, #Rank I use T4 GPU, so r=32 is too 'OOM' for me.
lora_alpha=16,
target_modules=["q_proj", "k_proj", "v_proj", "o_proj"],
bias="none",
lora_dropout=0.01, # Conventional
task_type="CAUSAL_LM",
)

# 1 - Enabling gradient checkpointing to reduce memory usage during fine-tuning
original_model.gradient_checkpointing_enable()
# 2 - Using the prepare_model_for_kbit_training method from PEFT
original_model = prepare_model_for_kbit_training(original_model)
peft_model = get_peft_model(original_model, config)

In [13]:
output_dir = '/content/peft-dialogue-summary-training/final-checkpoint'
import transformers
from transformers import TrainingArguments

peft_training_args = TrainingArguments(
output_dir = output_dir,
warmup_steps=1,
per_device_train_batch_size=1,
gradient_accumulation_steps=1,
max_steps=512, # 'OOM'problem, I have to reduce maz_steps
learning_rate=2e-4,
optim="paged_adamw_8bit",
logging_steps=100,
logging_dir="/content/logs",
save_strategy="steps",
save_steps=100,
eval_strategy="steps", #should replace evaluation_strategy="steps" with eval_strategy="steps". The argument name was changed in newer versions of the transformers library.
eval_steps=100,
do_eval=True,
gradient_checkpointing=True,
report_to="none",
overwrite_output_dir = True,
group_by_length=True,
#save_total_limit=2,             # 控制checkpoint数量，省空间
)

peft_model.config.use_cache = False

# Explicitly move the model to the GPU before initializing the trainer
#peft_model.to('cuda')#为什么不要 peft_model.to('cuda')

#对 4bit/8bit（bitsandbytes）+ device_map="auto" 的模型，手动 .to('cuda') 容易让部分量化层仍留在 CPU，导致 Accelerate 在比较设备时拿到 None（CPU 的 index），从而触发 torch.device(None) 的报错。
#让 from_pretrained(device_map="auto") 和 Trainer/Accelerate 统一管理设备放置最稳妥。


peft_trainer = transformers.Trainer(
model=peft_model,
train_dataset=train_dataset,
eval_dataset=eval_dataset,
args=peft_training_args,
data_collator=transformers.DataCollatorForLanguageModeling(tokenizer,
mlm=False),
)

Train peft model, and check how many parameters trained in training process.

Free memory after training because we only care trained new model.

In [14]:
peft_trainer.train()
# OutOfMemoryError OOM 多半是“序列太长 + CE上 fp32 中间张量占内存”。首选降低 max_length，并开启 fp16/bf16。
#训练先关评估，能跑通后再恢复。
#不手动 .to("cuda")，把设备交给 device_map="auto"。
#必要时把 LoRA rank 降到 8/16。









TrainOutput(global_step=512, training_loss=1.3809343315660954, metrics={'train_runtime': 395.6839, 'train_samples_per_second': 1.294, 'train_steps_per_second': 1.294, 'total_flos': 5977928817254400.0, 'train_loss': 1.3809343315660954, 'epoch': 0.256128064032016})

Since we encountered issues with the standard `Trainer`, let's try using the `SFTTrainer` from the `trl` library, which is designed for supervised fine-tuning tasks like this and might handle the integration with QLoRA and BitsAndBytes more smoothly.

In [15]:
'''This is a backup training plan if the standard Trainer above fails.
from trl import SFTTrainer

# Reusing the previously defined model, tokenizer, datasets, and training arguments

sft_trainer = SFTTrainer(
    model=peft_model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    peft_config=config, # Use the LoRA config here
    dataset_text_field="text", # Specify the column containing the formatted text
    max_seq_length=max_length,
    tokenizer=tokenizer,
    args=peft_training_args,
    packing=False,
)

# Start training with the SFTTrainer
sft_trainer.train()'''

'This is a backup training plan if the standard Trainer above fails.\nfrom trl import SFTTrainer\n\n# Reusing the previously defined model, tokenizer, datasets, and training arguments\n\nsft_trainer = SFTTrainer(\n    model=peft_model,\n    train_dataset=train_dataset,\n    eval_dataset=eval_dataset,\n    peft_config=config, # Use the LoRA config here\n    dataset_text_field="text", # Specify the column containing the formatted text\n    max_seq_length=max_length,\n    tokenizer=tokenizer,\n    args=peft_training_args,\n    packing=False,\n)\n\n# Start training with the SFTTrainer\nsft_trainer.train()'

After training, you can evaluate the model and save the trained adapter weights.

In [16]:
def print_number_of_trainable_model_parameters(model):
    trainable_model_params = 0
    all_model_params = 0
    for _, param in model.named_parameters():
        all_model_params += param.numel()
        if param.requires_grad:
            trainable_model_params += param.numel()
    return f"trainable model parameters: {trainable_model_params}\n \
all model parameters: {all_model_params}\n \
percentage of trainable model parameters: \
{100 * trainable_model_params / all_model_params:.2f}%"

# original_model is deleted before this function is called, so this will raise an error.
# I'll print the trainable parameters before deleting the original model.
print_number_of_trainable_model_parameters(peft_model)




'trainable model parameters: 13631488\n all model parameters: 4554231808\n percentage of trainable model parameters: 0.30%'

In [17]:
# Free memory（optional，I use colab pro and choose 80G ram, which is more than enough）
'''del original_model # original_model is not defined in this cell, so this will raise an error.
del peft_trainer
torch.cuda.empty_cache()'''

'del original_model # original_model is not defined in this cell, so this will raise an error.\ndel peft_trainer\ntorch.cuda.empty_cache()'

Now added trained Lora parameters \delta W to base model parameters W, and check by generating summary.

In [18]:
#import base model again if uncomment Free memory step.
compute_dtype = getattr(torch, "float16")
quant_config = BitsAndBytesConfig(
load_in_4bit=True,
bnb_4bit_quant_type="nf4",
bnb_4bit_compute_dtype=compute_dtype,
bnb_4bit_use_double_quant=True,
)
model_path = "meta-llama/Meta-Llama-3-8B-Instruct"
base_model = AutoModelForCausalLM.from_pretrained(
model_path,
dtype=compute_dtype,#`torch_dtype` is deprecated! Use `dtype` instead!
device_map={"": 0},
quantization_config=quant_config
)

Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]

In [19]:
#import tokenizer again if uncomment Free memory step.
eval_tokenizer = AutoTokenizer.from_pretrained(model_path, add_bos_token=True,
trust_remote_code=True, use_fast=False)
eval_tokenizer.pad_token = eval_tokenizer.eos_token

In [20]:
# Merge parameters from Lora and base model
from peft import PeftModel
ft_model = PeftModel.from_pretrained(
base_model,
"/content/peft-dialogue-summary-training/final-checkpoint/checkpoint-500",
torch_dtype=torch.float16,
is_trainable=False
) #作用：把“已经训练好的 LoRA 适配器权重”从磁盘（或Hub）加载到一个基座模型上，返回一个可用的 PeftModel。
#注意：from_pretrained 是“挂载适配器”，并没有把 ΔW 合并进基座权重。若你想得到“已融合”的单一模型，需要再调用 merge_and_unload()

In [21]:
%%time
index = 0
prompt = dataset['test'][index]['dialogue']
summary = dataset['test'][index]['summary']
formatted_prompt = f"Instruct: Summarize the following conversation.\nInput:{prompt}\nOutput:\n"
res = gen(ft_model,formatted_prompt,100,)
#print(res[0])
output = res[0].split('Output:\n')[1]
dash_line = '-'.join('' for x in range(200))
print(dash_line)
print(f'INPUT PROMPT:\n{formatted_prompt}')
print(dash_line)
print(f'BASELINE HUMAN SUMMARY:\n{summary}\n')
print(dash_line)
print(f'PEFT MODEL GENERATION:\n{output}')

# It's clear to see that fine-tuned ft_model has a summary close to human baseline summary from dataset.
# Compared with base model, ft_model has slightly better ability of dialog summary!
# I add 'scores' to evaluate how close the generated summary is to human baseline summary. Please check it on github

Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.


-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
INPUT PROMPT:
Instruct: Summarize the following conversation.
Input:#Person1#: Ms. Dawson, I need you to take a dictation for me.
#Person2#: Yes, sir...
#Person1#: This should go out as an intra-office memorandum to all employees by this afternoon. Are you ready?
#Person2#: Yes, sir. Go ahead.
#Person1#: Attention all staff... Effective immediately, all office communications are restricted to email correspondence and official memos. The use of Instant Message programs by employees during working hours is strictly prohibited.
#Person2#: Sir, does this apply to intra-office communications only? Or will it also restrict external communications?
#Person1#: It should apply to all communications, not only in this office between employees, but also any outside communications.
#Person2#: But sir, 

Next, based on ref summary from original dataset, we use Rouge score from HuggingFace to evaluate RougeL scores of base Llama 3 model and our finetuned model. We expect an improvment of scores.

In [22]:
#Prepare summary dataset from ref, Llama, finetuned model
ref_origin = []
for i in range(0,10):
  ref_origin.append(dataset['test'][i]['summary'])

pprint(ref_origin[0])

ft_summary = []
for i in range(0,10):
  index = i
  prompt = dataset['test'][index]['dialogue']
  summary = dataset['test'][index]['summary']
  formatted_prompt = f"Instruct: Summarize the following conversation.\nInput: {prompt}\nOutput:\n"
  #pprint(formatted_prompt)
  res = gen(ft_model,formatted_prompt,100,)
  dash_line = '-'.join('' for x in range(100))
  output = res[0].split('Output:\n')[1]
  ft_summary.append(output)

print(dash_line)
pprint(f'MODEL GENERATION - FEW-SHOT:\n{ft_summary[0]}')

Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.


('Ms. Dawson helps #Person1# to write a memo to inform every employee that '
 'they have to change the communication method and should not use Instant '
 'Messaging anymore.')


Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.


---------------------------------------------------------------------------------------------------
('MODEL GENERATION - FEW-SHOT:\n'
 'Ms. Dawson takes a dictation from #Person1# to restrict all office '
 'communications to email correspondence and official memos. #Person1# tells '
 'her the policy will apply to all communications, including internal and '
 'external ones. #Person1# also warns that employees who persist in using '
 'Instant Messaging will face termination. Ms. Dawson will get the memo typed '
 'up and distributed to all employees before 4 pm. #Person1# asks her to '
 'continue with the memo. Ms. Dawson writes down what')


#现象

问题：

第一次加载 original_model（刚从 from_pretrained 导入时） → 生成正常

训练完 PEFT（QLoRA）模型后

ft_model → 生成答案更好

同样的 original_model 再次生成 → 只输出 “Ms\n\n\n\n…”（退化输出）

回答 from GPT5:

🎯 最终原因：你的 original_model 被训练流程“破坏”了（状态被改写）

我已经完整阅读并逐行检查你整个训练流程。你在训练 QLoRA 的时候直接在 original_model 上注入了 LoRA 层、修改了模型 config、启用了 gradient checkpointing、禁用了 cache，并在 prepare_model_for_kbit_training() 里改变了模型 forward 行为。

🔥🔥 所以 original_model 已经不再是原始模型了！

你训练完之后：

original_model = prepare_model_for_kbit_training(original_model)
peft_model = get_peft_model(original_model, config)


这段操作直接在 原模型内部做了以下更改（这些都是 peft + QLoRA 会干的）：

📌 1. 你把 LoRA adapter 注入了 original_model 自己

PEFT 的实现方式是：

把 LoRA module 注入到 base model 的 Q/K/V/O projection 层

替换了这些 Linear 层的 forward 函数

修改了 model.config 里的 use_cache、gradient_checkpointing、training flags

所以：

original_model 在你训练完之后已经不是原来的 8B 模型，它已经是“QLoRA 训练准备状态下的模型”，forward 路径不一样。

📌 2. 你对 original_model 开启了 gradient checkpointing

来自 PDF：

original_model.gradient_checkpointing_enable()


这会改变模型前向传播的行为，尤其会影响：

KV Cache

hidden states 顺序

dropout

training flag

对于推理来说，这是不合法状态。

于是 original_model 在推理阶段会输出垃圾，例如：

Ms
\n\n\n\n...

📌 3. prepare_model_for_kbit_training() 直接修改了模型内部

该函数会：

将 layernorm 权重转成 fp32

修改 model.forward，让某些层返回新的结构

将 embedding / lm_head cast 成不同 dtype

注入 zero_grad hooks

禁用/修改缓存行为（特别是 use_cache=False 时）

它直接改变模型行为。

📌 4. 你在训练前把 original_model.config.use_cache 改成 False

PDF 原文：

peft_model.config.use_cache = False


由于 peft_model 是基于 original_model 构建的，
这实际上把 original_model.config.use_cache 也改成 False。

Llama 在推理时 use_cache=False 会导致重复生成换行、状态错乱、退化输出。

这也是你看到：

Ms
\n\n\n\n...


的直接原因之一。

📌 5. 你在训练过程中修改了 original_model 对象本身

关键问题：
你没有在训练前 clone 一份模型。

训练流程里用的是：

original_model  # <-- 训练前的模型
prepare_model_for_kbit_training(original_model)
peft_model = get_peft_model(original_model)


也就是说：

training 是在 original_model 上“原地”修改的。

训练后的 original_model = 被改写、注入 LoRA、禁用 cache、修改 forward、开启 checkpointing 的模型

它不能再做正常推理。

📌 6. 你训练后 “重建 base_model” 的方式是正确的，但你没有重建 original_model

你训练完之后做了：

base_model = AutoModelForCausalLM.from_pretrained(...)


这是正确的。

但是你继续用 old original_model（被训练破坏过的模型）来做推理，就会输出垃圾。

🧨 所以最终现象就产生了：
模型	状态	输出
original_model（第一次加载）	干净	正常 summary
original_model（训练后）	已被 QLoRA 修改，forward 改变，cache 禁用	Ms + 大量空行（退化输出）
ft_model	正常加载（合并 LoRA）	输出良好
🎯 解决方案（最推荐）
✔ 解决方案 1：训练后重新加载 original_model

不要再使用被训练过的 original_model。

训练后：

compute_dtype = getattr(torch, "float16")
quant_config = BitsAndBytesConfig(
  load_in_4bit=True,
  bnb_4bit_quant_type="nf4",
  bnb_4bit_compute_dtype=compute_dtype,
  bnb_4bit_use_double_quant=True,
)

original_model = AutoModelForCausalLM.from_pretrained(
    "meta-llama/Meta-Llama-3-8B-Instruct",
    dtype=compute_dtype,
    device_map="auto",
    quantization_config=quant_config
)


然后 original_model 会再次恢复正常输出。

✔ 解决方案 2：训练前 clone 模型
import copy
base_model = copy.deepcopy(original_model)


让 original_model 保持未被污染。

✔ 解决方案 3（次优）

把你推理 prompt 改为 Llama3 官方 ChatML，避免 Ms \n\n\n\n

🎯 最终一句话总结

**因为你在训练 LoRA 的过程中修改了 original_model（注入 LoRA、改变 forward、禁用 use_cache、开启 checkpointing），导致 original_model 在训练后不能再正常推理，输出变成 “Ms + 一堆换行”。

重新加载 original_model 就恢复正常。**

In [ ]:
!pip install evaluate
!pip install rouge_score
import evaluate

In [25]:
#Compute RougeL scores of finetuned and Llama model. Check the differnece


# 加载 ROUGE metric
rouge = evaluate.load("rouge")

# 计算整体 ROUGE 分数
scores_ft = rouge.compute(
    predictions=ft_summary,
    references=ref_origin,
    # 可选参数：
    use_stemmer=True  # 英文可以开，对中文一般无意义
)

print(scores_ft)

scores_noft = rouge.compute(
    predictions=base_summary,
    references=ref_origin,
    # 可选参数：
    use_stemmer=True  # 英文可以开，对中文一般无意义
)

print(scores_noft)

  Preparing metadata (setup.py) ... done
  Created wheel for rouge_score: filename=rouge_score-0.1.2-py3-none-any.whl size=24934 sha256=5ef24849223a1c3399cc2424196ed4412f713c80712da8b1e8a6cac8dffcf352
  Stored in directory: /root/.cache/pip/wheels/85/9d/af/01feefbe7d55ef5468796f0c68225b6788e85d9d0a281e7a70
Successfully built rouge_score


{'rouge1': np.float64(0.32102673459778264), 'rouge2': np.float64(0.14050694774228806), 'rougeL': np.float64(0.23785597432305772), 'rougeLsum': np.float64(0.23747036434151342)}
{'rouge1': np.float64(0.25983521043970326), 'rouge2': np.float64(0.08840670723749743), 'rougeL': np.float64(0.1785822714906734), 'rougeLsum': np.float64(0.17853018443207122)}


##Conclusion

ft_model{'rouge1': np.float64(0.32102673459778264), 'rouge2': np.float64(0.14050694774228806), 'rougeL': np.float64(0.23785597432305772), 'rougeLsum': np.float64(0.23747036434151342)}

base_model{'rouge1': np.float64(0.25983521043970326), 'rouge2': np.float64(0.08840670723749743), 'rougeL': np.float64(0.1785822714906734), 'rougeLsum': np.float64(0.17853018443207122)}

Therefore, ft_model > base_model, has 30% better ability for summerizing.

## Simple induction of Rouge Score

ROUGE = Recall-Oriented Understudy for Gisting Evaluation
是自动摘要任务中最经典的评估指标。它本质上测量：
模型生成的 summary 中，有多少内容和参考摘要（人工写的）重叠？

| 指标             | 关注点                            |
| -------------- | ------------------------------ |
| **ROUGE-1**    | unigram（1 元词/字）重叠              |
| **ROUGE-2**    | bigram（2 元连续词）重叠               |
| **ROUGE-L**    | 最长公共子序列（LCS），关注顺序一致性           |
| **ROUGE-Lsum** | sentence-wise LCS（专门优化新闻/摘要任务） |

定义

# ROUGE 指标公式汇总（ROUGE-1 / ROUGE-2 / ROUGE-L / ROUGE-Lsum）

下面设：

- 参考摘要（reference）记为序列  
  \( R = (r_1, r_2, \dots, r_m) \)
- 模型摘要（prediction）记为序列  
  \( P = (p_1, p_2, \dots, p_n) \)

这里的 \(r_i, p_j\) 可以理解为 **英文的词（token）**。

---

## 1️⃣ ROUGE-1（unigram 重叠）

ROUGE-1 衡量的是 **unigram（单个词）重叠率**，一般用召回率（recall）形式。

定义词的 multiset（多重集合）：

- \( U(R) \)：参考摘要的所有 unigram 及其出现次数  
- \( U(P) \)：预测摘要的所有 unigram 及其出现次数  

那么所有 unigram 的重叠数为：

\[
\text{Overlap}_1 = \sum_{w \in V} \min\big( \text{count}_R(w), \ \text{count}_P(w) \big)
\]

其中 \(V\) 是所有可能词的集合。

**ROUGE-1（召回）** 定义为：

\[
\text{ROUGE-1} =
\frac{\text{Overlap}_1}{\sum_{w \in V} \text{count}_R(w)}
= \frac{\text{Overlap}_1}{|R|}
\]

即：  
> 预测摘要中命中参考摘要的 unigram 数量 ÷ 参考摘要总 unigram 数量。

---

## 2️⃣ ROUGE-2（bigram 重叠）

ROUGE-2 衡量的是 **bigram（连续两词）重叠率**。  
先定义 bigram 序列：

- 参考摘要的 bigram 序列：  
  \[
  B(R) = \big\{(r_1,r_2), (r_2,r_3), \dots, (r_{m-1},r_m)\big\}
  \]
- 预测摘要的 bigram 序列：  
  \[
  B(P) = \big\{(p_1,p_2), (p_2,p_3), \dots, (p_{n-1},p_n)\big\}
  \]

同样用 multiset 计数，用：

\[
\text{Overlap}_2 = \sum_{b \in \mathcal{B}}
\min\big( \text{count}_{B(R)}(b), \ \text{count}_{B(P)}(b) \big)
\]

其中 \(\mathcal{B}\) 是所有可能 bigram 的集合。

**ROUGE-2（召回）** 为：

\[
\text{ROUGE-2} =
\frac{\text{Overlap}_2}{|B(R)|}
=
\frac{\text{Overlap}_2}{m-1}
\]

即：  
> 预测摘要中命中参考摘要的 bigram 数量 ÷ 参考摘要总 bigram 数量。

---

## 3️⃣ ROUGE-L（基于最长公共子序列 LCS）

ROUGE-L 使用 **Longest Common Subsequence（LCS，最长公共子序列）** 衡量两段文本的 **顺序一致性**。

给定序列 \(R, P\)，其 LCS 长度记为：

\[
LCS(R, P)
\]

LCS 是在保持顺序的前提下，找到两个序列中最长的“相同子序列”，不要求连续。

然后定义基于 LCS 的召回和精度：

\[
R_L = \frac{LCS(R,P)}{|R|}
\]

\[
P_L = \frac{LCS(R,P)}{|P|}
\]

再定义 F1（或 F\(_\beta\)）：

\[
F_L =
\frac{(1+\beta^2)\,R_L\,P_L}{R_L + P_L}
\]

在很多实现中，\(\beta = 1\)，即普通 F1：

\[
F_L = \frac{2\,R_L\,P_L}{R_L + P_L}
\]

**实践中常报告的 ROUGE-L 就是这个 \(F_L\)**。

---

## 4️⃣ ROUGE-Lsum（面向多句摘要的 ROUGE-L）

ROUGE-L 对整段文本当作一个长序列算一次 LCS，适合 **短文本**。  
但真实摘要（尤其是新闻、对话摘要）往往由多句组成，句子级结构也很重要。

**ROUGE-Lsum** 的思想：

> 把参考摘要和预测摘要都切成多个句子，  
> 然后按“句子对”匹配 LCS，再对所有句子求一个整体的 LCS 召回 / 精度。

### 4.1 句子级表示

把参考摘要拆成 \(m\) 个句子：

\[
R = \{ s^R_1, s^R_2, \dots, s^R_m \}
\]

把预测摘要拆成 \(n\) 个句子：

\[
P = \{ s^P_1, s^P_2, \dots, s^P_n \}
\]

每个句子本身又是一个 token 序列。

### 4.2 “一对多句子匹配”策略

对参考摘要中的每个句子 \(s^R_i\)，  
在预测摘要所有句子中，找到 **LCS 最长的那一个句子**：

\[
LCS_i = \max_{j} LCS\big(s^R_i,\ s^P_j\big)
\]

也就是：  
> 对每个参考句子，只找一个最相似的预测句子，不强行一一配对所有句子。

然后对所有参考句子，把这些 LCS 长度加起来：

\[
\text{LCS}_{\text{sum}} = \sum_{i=1}^{m} LCS_i
\]

这就是“句子级 LCS 的总和”。

### 4.3 ROUGE-Lsum 的召回和精度

定义：

- 参考摘要总 token 数量：  
  \[
  |R| = \sum_{i=1}^{m} \big|s^R_i\big|
  \]
- 预测摘要总 token 数量：  
  \[
  |P| = \sum_{j=1}^{n} \big|s^P_j\big|
  \]

则 ROUGE-Lsum 的 recall 和 precision 为：

\[
R_{Lsum} = \frac{\text{LCS}_{\text{sum}}}{|R|}
\]

\[
P_{Lsum} = \frac{\text{LCS}_{\text{sum}}}{|P|}
\]

同样构造 F1：

\[
F_{Lsum} =
\frac{(1+\beta^2)\,R_{Lsum}\,P_{Lsum}}{R_{Lsum} + P_{Lsum}}
\]

通常 \(\beta = 1\) 时：

\[
F_{Lsum} = \frac{2\,R_{Lsum}\,P_{Lsum}}{R_{Lsum} + P_{Lsum}}
\]

**实践中报告的 ROUGE-Lsum 通常就是这个 \(F_{Lsum}\)**。

---

## 4.4 为什么需要 ROUGE-Lsum？

对摘要/新闻这类任务：

- 一篇摘要常由多个句子组成  
- 句子内部的顺序，以及句子和句子之间的结构都很重要  
- 简单地把整篇拼成一个长序列算 LCS，可能会：
  - 把跨句子的匹配算在一起，丢失句子边界的信息
  - 让长句子完全主导得分

ROUGE-Lsum 通过 **“句子级一对多匹配 + 汇总 LCS”** 的方式：

- 更关注“参考摘要每个句子有没有被覆盖到”  
- 更贴合人类对“摘要是否完整覆盖关键信息”的直觉  
- 适合评估 multi-sentence summary（这也是 Hugging Face 默认给 `rougeLsum` 的原因）

---
	​


注意：
Hugging Face 的 ROUGE 默认是 recall-based。
